roachtest: schemachange/index/tpcc/w=100 failed

SHA: https://github.com/cockroachdb/cockroach/commits/dfa23c01e4ea39b19ca8b2e5c8a4e7cf9b9445f4

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/index/tpcc/w=100 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1189954&tab=buildLog

```
The test failed on master:
	test.go:1202: test timed out (30m0s)
	cluster.go:1267,tpcc.go:142,schemachange.go:310,test.go:1214: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1189954-schemachange-index-tpcc-w-100:5 -- ./workload check tpcc --warehouses=100 {pgurl:1} returned:
		stderr:
		
		stdout:
		I190321 13:44:27.112085 1 workload/tpcc/tpcc.go:292  check 3.3.2.1 took 9.626722644s
		I190321 13:45:00.489455 1 workload/tpcc/tpcc.go:292  check 3.3.2.2 took 33.377279093s
		I190321 13:45:11.828237 1 workload/tpcc/tpcc.go:292  check 3.3.2.3 took 11.338718794s
		I190321 13:47:44.111289 1 workload/tpcc/tpcc.go:292  check 3.3.2.4 took 2m32.282972385s
		I190321 13:47:50.255716 1 workload/tpcc/tpcc.go:292  check 3.3.2.5 took 6.144344078s
		I190321 13:48:12.868170 1 workload/tpcc/tpcc.go:292  check 3.3.2.7 took 22.612378595s
		I190321 13:48:15.409456 1 workload/tpcc/tpcc.go:292  check 3.3.2.8 took 2.541209195s
		: signal: killed

```

SHA: https://github.com/cockroachdb/cockroach/commits/5f358ed804af05f8c4b404efc4d8a282d8e0916c

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/index/tpcc/w=100 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1361643&tab=buildLog

```
The test failed on branch=master, cloud=gce:
	cluster.go:1511,tpcc.go:167,cluster.go:1849,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1361643-schemachange-index-tpcc-w-100:5 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=15m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		    8.3      0.0      0.0      0.0      0.0 delivery
		  12m47s   177934            0.0           82.6      0.0      0.0      0.0      0.0 newOrder
		  12m47s   177934            0.0            8.3      0.0      0.0      0.0      0.0 orderStatus
		  12m47s   177934            0.0           82.2      0.0      0.0      0.0      0.0 payment
		  12m47s   177934            0.0            8.3      0.0      0.0      0.0      0.0 stockLevel
		E190627 02:18:30.729951 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.55:26257: connect: connection refused
		  12m48s   233955            0.0            8.3      0.0      0.0      0.0      0.0 delivery
		  12m48s   233955            0.0           82.5      0.0      0.0      0.0      0.0 newOrder
		  12m48s   233955            0.0            8.3      0.0      0.0      0.0      0.0 orderStatus
		  12m48s   233955            0.0           82.1      0.0      0.0      0.0      0.0 payment
		  12m48s   233955            0.0            8.3      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: unexpected node event: 3: dead
	cluster.go:1033,context.go:122,cluster.go:1022,asm_amd64.s:522,panic.go:397,test.go:783,test.go:769,cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1361643-schemachange-index-tpcc-w-100 --oneshot --ignore-empty-nodes: exit status 1 2: 3561
		4: 3328
		3: dead
		1: 3420
		5: skipped
		Error:  3: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/90841a6559df9d9a4724e1d30490951bbdb811b4

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/index/tpcc/w=100 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1364443&tab=buildLog

```
The test failed on branch=provisional_201906271846_v19.2.0-alpha.20190701, cloud=gce:
	cluster.go:1511,tpcc.go:167,cluster.go:1849,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1364443-schemachange-index-tpcc-w-100:5 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=15m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		un.go:427  error in newOrder: dial tcp 10.128.0.58:26257: connect: connection refused
		   1m10s    27330            1.0            7.9    104.9    104.9    104.9    104.9 delivery
		   1m10s    27330            0.0           75.0      0.0      0.0      0.0      0.0 newOrder
		   1m10s    27330            0.0            7.7      0.0      0.0      0.0      0.0 orderStatus
		   1m10s    27330            6.0           72.7   1811.9  11811.2  11811.2  11811.2 payment
		   1m10s    27330            0.0            7.9      0.0      0.0      0.0      0.0 stockLevel
		   1m11s    27330            0.0            7.8      0.0      0.0      0.0      0.0 delivery
		   1m11s    27330            0.0           73.9      0.0      0.0      0.0      0.0 newOrder
		   1m11s    27330            0.0            7.6      0.0      0.0      0.0      0.0 orderStatus
		   1m11s    27330            0.0           71.7      0.0      0.0      0.0      0.0 payment
		   1m11s    27330            0.0            7.8      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: unexpected node event: 2: dead
	cluster.go:1033,context.go:122,cluster.go:1022,panic.go:406,test.go:783,test.go:769,cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1364443-schemachange-index-tpcc-w-100 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		4: 3407
		1: 3509
		2: dead
		3: 3877
		Error:  2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/537767ac9daa52b0026bb957d7010e3b88b61071

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/index/tpcc/w=100 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1364821&tab=buildLog

```
The test failed on branch=master, cloud=gce:
	cluster.go:1511,tpcc.go:167,cluster.go:1849,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1364821-schemachange-index-tpcc-w-100:5 -- ./workload run tpcc --warehouses=100 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=15m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		0      0.0      0.0 orderStatus
		  10m52s   113290            0.0           84.3      0.0      0.0      0.0      0.0 payment
		  10m52s   113290            0.0            8.5      0.0      0.0      0.0      0.0 stockLevel
		E190628 19:33:35.310787 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.58:26257: connect: connection refused
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		  10m53s   153374            0.0            8.5      0.0      0.0      0.0      0.0 delivery
		  10m53s   153374            0.0           84.5      0.0      0.0      0.0      0.0 newOrder
		  10m53s   153374            0.0            8.5      0.0      0.0      0.0      0.0 orderStatus
		  10m53s   153374            0.0           84.1      0.0      0.0      0.0      0.0 payment
		  10m53s   153374            0.0            8.4      0.0      0.0      0.0      0.0 stockLevel
		E190628 19:33:36.310829 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.58:26257: connect: connection refused
		: signal: killed
	cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: unexpected node event: 3: dead
	cluster.go:1033,context.go:122,cluster.go:1022,panic.go:406,test.go:783,test.go:769,cluster.go:1870,tpcc.go:177,schemachange.go:300,test.go:1249: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1364821-schemachange-index-tpcc-w-100 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		2: 3376
		3: dead
		4: 3412
		1: 4102
		Error:  3: dead

```

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -1029,6 +1029,17 @@ func maybeSwapErrorIndex(pErr *roachpb.Error, a, b int) {
 	}
 }
 
+// mergeErrors merges the two errors, combining their transaction state and
+// returning the error with the highest priority.
+func mergeErrors(pErr1, pErr2 *roachpb.Error) *roachpb.Error {
+	ret, drop := pErr1, pErr2
+	if roachpb.ErrPriority(drop.GoError()) > roachpb.ErrPriority(ret.GoError()) {
+		ret, drop = drop, ret
+	}
+	ret.UpdateTxn(drop.GetTxn())
+	return ret
+}
+
 // divideAndSendBatchToRanges sends the supplied batch to all of the
 // ranges which comprise the span specified by rs. The batch request
 // is trimmed against each range which is part of the span and sent
@@ -1139,10 +1150,10 @@ func (ds *DistSender) divideAndSendBatchToRanges(
 					// the batch was parallelized and part of it succeeded.
 					pErr.UpdateTxn(br.Txn)
 				} else {
-					// Even though we ignore the second error, update the first
-					// error's transaction with any new information from the
-					// second error. This may contain interesting updates.
-					pErr.UpdateTxn(resp.pErr.GetTxn())
+					// The batch was split and saw (at least) two different errors.
+					// Merge their transaction state and determine which to return
+					// based on their priorities.
+					pErr = mergeErrors(pErr, resp.pErr)
 				}
 				continue
 			}

--- a/pkg/kv/dist_sender_test.go
+++ b/pkg/kv/dist_sender_test.go
@@ -2638,6 +2638,161 @@ func TestGatewayNodeID(t *testing.T) {
 	}
 }
 
+// TestMultipleErrorsMerged tests that DistSender prioritizes errors that are
+// returned from concurrent partial batches and returns the "best" one after
+// merging the transaction metadata passed on the errors.
+func TestMultipleErrorsMerged(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.TODO())
+
+	clock := hlc.NewClock(hlc.UnixNano, time.Nanosecond)
+	rpcContext := rpc.NewInsecureTestingContext(clock, stopper)
+	g := makeGossip(t, stopper, rpcContext)
+
+	if err := g.SetNodeDescriptor(newNodeDesc(1)); err != nil {
+		t.Fatal(err)
+	}
+	nd := &roachpb.NodeDescriptor{
+		NodeID:  roachpb.NodeID(1),
+		Address: util.MakeUnresolvedAddr(testAddress.Network(), testAddress.String()),
+	}
+	if err := g.AddInfoProto(gossip.MakeNodeIDKey(roachpb.NodeID(1)), nd, time.Hour); err != nil {
+		t.Fatal(err)
+	}
+
+	// Fill MockRangeDescriptorDB with two descriptors.
+	var descriptor1 = roachpb.RangeDescriptor{
+		RangeID:  2,
+		StartKey: testMetaEndKey,
+		EndKey:   roachpb.RKey("b"),
+		InternalReplicas: []roachpb.ReplicaDescriptor{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	var descriptor2 = roachpb.RangeDescriptor{
+		RangeID:  3,
+		StartKey: roachpb.RKey("b"),
+		EndKey:   roachpb.RKeyMax,
+		InternalReplicas: []roachpb.ReplicaDescriptor{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	descDB := mockRangeDescriptorDBForDescs(
+		testMetaRangeDescriptor,
+		descriptor1,
+		descriptor2,
+	)
+
+	txn := roachpb.MakeTransaction(
+		"test", nil /* baseKey */, roachpb.NormalUserPriority,
+		clock.Now(), clock.MaxOffset().Nanoseconds(),
+	)
+	retryErr := roachpb.NewTransactionRetryError(roachpb.RETRY_SERIALIZABLE, "test err")
+	abortErr := roachpb.NewTransactionAbortedError(roachpb.ABORT_REASON_ABORTED_RECORD_FOUND)
+	conditionFailedErr := &roachpb.ConditionFailedError{}
+
+	testCases := []struct {
+		err1, err2 error
+		expErr     string
+	}{
+		{
+			err1:   retryErr,
+			err2:   nil,
+			expErr: "mixed success and failure: TransactionRetryError: retry txn (RETRY_SERIALIZABLE)",
+		},
+		{
+			err1:   abortErr,
+			err2:   nil,
+			expErr: "mixed success and failure: TransactionAbortedError(ABORT_REASON_ABORTED_RECORD_FOUND)",
+		},
+		{
+			err1:   conditionFailedErr,
+			err2:   nil,
+			expErr: "mixed success and failure: unexpected value",
+		},
+		{
+			err1:   retryErr,
+			err2:   retryErr,
+			expErr: "TransactionRetryError: retry txn (RETRY_SERIALIZABLE)",
+		},
+		{
+			err1:   retryErr,
+			err2:   abortErr,
+			expErr: "TransactionAbortedError(ABORT_REASON_ABORTED_RECORD_FOUND)",
+		},
+		{
+			err1:   abortErr,
+			err2:   abortErr,
+			expErr: "TransactionAbortedError(ABORT_REASON_ABORTED_RECORD_FOUND)",
+		},
+		{
+			err1:   retryErr,
+			err2:   conditionFailedErr,
+			expErr: "unexpected value",
+		},
+		{
+			err1:   abortErr,
+			err2:   conditionFailedErr,
+			expErr: "unexpected value",
+		},
+		{
+			err1:   conditionFailedErr,
+			err2:   conditionFailedErr,
+			expErr: "unexpected value",
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			var testFn simpleSendFn = func(
+				_ context.Context,
+				_ SendOptions,
+				_ ReplicaSlice,
+				ba roachpb.BatchRequest,
+			) (*roachpb.BatchResponse, error) {
+				reply := ba.CreateReply()
+				if delRng := ba.Requests[0].GetDeleteRange(); delRng == nil {
+					return nil, errors.Errorf("expected DeleteRange request, found %v", ba.Requests[0])
+				} else if delRng.Key.Equal(roachpb.Key("a")) {
+					reply.Error = roachpb.NewError(tc.err1)
+				} else if delRng.Key.Equal(roachpb.Key("b")) {
+					reply.Error = roachpb.NewError(tc.err2)
+				} else {
+					return nil, errors.Errorf("unexpected DeleteRange boundaries")
+				}
+				return reply, nil
+			}
+
+			cfg := DistSenderConfig{
+				AmbientCtx: log.AmbientContext{Tracer: tracing.NewTracer()},
+				Clock:      clock,
+				RPCContext: rpcContext,
+				TestingKnobs: ClientTestingKnobs{
+					TransportFactory: adaptSimpleTransport(testFn),
+				},
+				RangeDescriptorDB: descDB,
+			}
+			ds := NewDistSender(cfg, g)
+
+			var ba roachpb.BatchRequest
+			ba.Txn = txn.Clone()
+			ba.Add(roachpb.NewDeleteRange(roachpb.Key("a"), roachpb.Key("c"), false))
+
+			if _, pErr := ds.Send(context.Background(), ba); pErr == nil {
+				t.Fatalf("expected an error to be returned from distSender")
+			} else if !testutils.IsPError(pErr, regexp.QuoteMeta(tc.expErr)) {
+				t.Fatalf("expected error %q; found %v", tc.expErr, pErr)
+			}
+		})
+	}
+}
+
 // Regression test for #20067.
 // If a batch is partitioned into multiple partial batches, the
 // roachpb.Error.Index of each batch should correspond to its original index in
@@ -2660,7 +2815,6 @@ func TestErrorIndexAlignment(t *testing.T) {
 	}
 	if err := g.AddInfoProto(gossip.MakeNodeIDKey(roachpb.NodeID(1)), nd, time.Hour); err != nil {
 		t.Fatal(err)
-
 	}
 
 	// Fill MockRangeDescriptorDB with two descriptors.

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -1217,6 +1217,9 @@ func PrepareTransactionForRetry(
 		log.Fatalf(ctx, "invalid retryable err (%T): %s", pErr.GetDetail(), pErr)
 	}
 	if !aborted {
+		if txn.Status.IsFinalized() {
+			log.Fatalf(ctx, "transaction unexpectedly finalized in (%T): %s", pErr.GetDetail(), pErr)
+		}
 		txn.Restart(pri, txn.Priority, txn.Timestamp)
 	}
 	return txn

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -39,16 +39,56 @@ func (e *UnhandledRetryableError) Error() string {
 
 var _ error = &UnhandledRetryableError{}
 
+// transactionRestartError is an interface implemented by errors that cause
+// a transaction to be restarted.
+type transactionRestartError interface {
+	canRestartTransaction() TransactionRestart
+}
+
 // ErrorUnexpectedlySet creates a string to panic with when a response (typically
 // a roachpb.BatchResponse) unexpectedly has Error set in its response header.
 func ErrorUnexpectedlySet(culprit, response interface{}) string {
 	return fmt.Sprintf("error is unexpectedly set, culprit is %T:\n%+v", culprit, response)
 }
 
-// transactionRestartError is an interface implemented by errors that cause
-// a transaction to be restarted.
-type transactionRestartError interface {
-	canRestartTransaction() TransactionRestart
+// ErrorPriority is used to rank errors such that the "best" one is chosen to be
+// presented as the batch result when a batch is split up and observes multiple
+// errors. Higher values correspond to higher priorities.
+type ErrorPriority int
+
+const (
+	_ ErrorPriority = iota
+	// ErrorScoreTxnRestart indicates that the transaction should be restarted
+	// with an incremented epoch.
+	ErrorScoreTxnRestart
+	// ErrorScoreTxnAbort indicates that the transaction is aborted. The
+	// operation can only try again under the purview of a new transaction.
+	ErrorScoreTxnAbort
+	// ErrorScoreNonRetriable indicates that the transaction performed an
+	// operation that does not warrant a retry. Often this indicates that the
+	// operation ran into a logic error. The error should be propagated to the
+	// client and the transaction should terminate immediately.
+	ErrorScoreNonRetriable
+)
+
+// ErrPriority computes the priority of the given error.
+func ErrPriority(err error) ErrorPriority {
+	if err == nil {
+		return 0
+	}
+	switch v := err.(type) {
+	case *UnhandledRetryableError:
+		if _, ok := v.PErr.GetDetail().(*TransactionAbortedError); ok {
+			return ErrorScoreTxnAbort
+		}
+		return ErrorScoreTxnRestart
+	case *TransactionRetryWithProtoRefreshError:
+		if v.PrevTxnAborted() {
+			return ErrorScoreTxnAbort
+		}
+		return ErrorScoreTxnRestart
+	}
+	return ErrorScoreNonRetriable
 }
 
 // NewError creates an Error from the given error.

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -306,17 +306,6 @@ func (dsp *DistSQLPlanner) Run(
 	flow.Cleanup(ctx)
 }
 
-// errorPriority is used to rank errors such that the "best" one is chosen to be
-// presented as the query result.
-type errorPriority int
-
-const (
-	scoreNoError errorPriority = iota
-	scoreTxnRestart
-	scoreTxnAbort
-	scoreNonRetriable
-)
-
 // DistSQLReceiver is a RowReceiver that writes results to a rowResultWriter.
 // This is where the DistSQL execution meets the SQL Session - the RowContainer
 // comes from a client Session.
@@ -534,7 +523,7 @@ func (r *DistSQLReceiver) Push(
 		if meta.Err != nil {
 			// Check if the error we just received should take precedence over a
 			// previous error (if any).
-			if errPriority(meta.Err) > errPriority(r.resultWriter.Err()) {
+			if roachpb.ErrPriority(meta.Err) > roachpb.ErrPriority(r.resultWriter.Err()) {
 				if r.txn != nil {
 					if retryErr, ok := meta.Err.(*roachpb.UnhandledRetryableError); ok {
 						// Update the txn in response to remote errors. In the non-DistSQL
@@ -640,30 +629,6 @@ func (r *DistSQLReceiver) Push(
 		return r.status
 	}
 	return r.status
-}
-
-// errPriority computes the priority of err.
-func errPriority(err error) errorPriority {
-	if err == nil {
-		return scoreNoError
-	}
-	err = errors.Cause(err)
-	if retryErr, ok := err.(*roachpb.UnhandledRetryableError); ok {
-		pErr := retryErr.PErr
-		switch pErr.GetDetail().(type) {
-		case *roachpb.TransactionAbortedError:
-			return scoreTxnAbort
-		default:
-			return scoreTxnRestart
-		}
-	}
-	if retryErr, ok := err.(*roachpb.TransactionRetryWithProtoRefreshError); ok {
-		if retryErr.PrevTxnAborted() {
-			return scoreTxnAbort
-		}
-		return scoreTxnRestart
-	}
-	return scoreNonRetriable
 }
 
 // ProducerDone is part of the RowReceiver interface.

--- a/pkg/storage/replica_raft.go
+++ b/pkg/storage/replica_raft.go
@@ -267,6 +267,10 @@ func (r *Replica) propose(ctx context.Context, p *ProposalData) (_ int64, pErr *
 		log.VEvent(p.ctx, 4, "sideloadable proposal detected")
 		version = raftVersionSideloaded
 		r.store.metrics.AddSSTableProposals.Inc(1)
+
+		if p.command.ReplicatedEvalResult.AddSSTable.Data == nil {
+			return 0, roachpb.NewErrorf("cannot sideload empty SSTable")
+		}
 	} else if log.V(4) {
 		log.Infof(p.ctx, "proposing command %x: %s", p.idKey, p.Request.Summary())
 	}

